File: athly-ios/AthlyRunner/Services/SplitCalculator.swift
import Foundation
import CoreLocation

/// Um split de 1 km (o último pode ser parcial). `paceSecondsPerKm` é derivado da distância
/// real do segmento — então o km parcial final também sai com pace correto.
struct KmSplit: Sendable {
    let kilometer: Int
    let startDate: Date
    let endDate: Date
    let distanceMeters: Double
    let durationSeconds: Double
    let elevationDelta: Double

    var paceSecondsPerKm: Double {
        distanceMeters > 0 ? durationSeconds / (distanceMeters / 1000.0) : 0
    }
}

/// Algoritmo único de splits por km, compartilhado entre o tracker ao vivo (`RunTracker`) e a
/// leitura de rota do Apple Health (`WorkoutDetailFetcher`). Centralizar evita que os dois
/// caminhos voltem a divergir — foi justamente a divergência entre eles que produziu o bug do
/// km 1 inflado (pace do split ~7' enquanto o pace ao vivo era 5'40-6').
///
/// O pace do split é calculado na MESMA base do tracker ao vivo:
/// - saltos de GPS rejeitados por plausibilidade de velocidade (≤ `maxPlausibleSpeed`);
/// - tempo parado / buracos de GPS (pausa, tela bloqueada) excluídos da duração;
/// - o relógio do km 1 só começa no primeiro movimento real (descarta o tempo parado inicial);
/// - fronteiras de km exatas por interpolação linear.
enum SplitCalculator {
    /// Intervalo de pausa explícita (botão de pausa). Cruzar um intervalo desses não conta nem
    /// distância (a corda entre fim e retomada é deriva sem sentido) nem tempo — espelha o
    /// caminho ao vivo, onde os fixes durante a pausa nem chegam a ser gravados.
    struct PauseInterval: Codable, Sendable, Equatable {
        let start: Date
        let end: Date
    }

    /// Velocidade máxima plausível em corrida (m/s ≈ 2:23/km). Acima disso é salto de GPS.
    static let maxPlausibleSpeed: Double = 7.0
    /// Gap (s) entre fixes acima do qual o intervalo é candidato a pausa/blackout de GPS.
    static let gapCapSeconds: Double = 6.0
    /// Velocidade (m/s) abaixo da qual um gap grande é tratado como parado (não conta tempo).
    /// Acima disso, um gap grande foi corrido de fato (ex.: blackout de GPS) e o tempo conta.
    static let stationarySpeed: Double = 0.5
    /// Distância (m) que marca o "primeiro movimento": o relógio do km 1 só começa aqui,
    /// descartando o tempo parado / GPS esquentando no início.
    static let startMovementMeters: Double = 20.0
    /// Distância mínima (m) para registrar um km parcial final.
    static let minTrailingMeters: Double = 50.0
    /// Gap máximo (s) que ainda recebe crédito de distância pela velocidade pré-gap.
    static let maxBridgeGapSeconds: Double = 90.0
    /// Janela (s de tempo em movimento) usada para estimar a velocidade pré-gap.
    static let speedLookbackSeconds: Double = 30.0

    /// Amostra normalizada da trajetória. `movingTime` é o tempo de movimento acumulado
    /// (pausas/buracos já descontados); `distance` é a distância acumulada já filtrada.
    private struct Sample {
        let date: Date
        let movingTime: Double
        let distance: Double
        let altitude: Double
    }

    /// Quebra os locations em splits de 1 km. Retorna `[]` se não houver dados suficientes.
    static func kmSplits(from locations: [CLLocation], pauses: [PauseInterval] = []) -> [KmSplit] {
        let samples = normalize(locations, pauses: pauses)
        guard samples.count >= 2 else { return [] }

        // Âncora de primeiro movimento: tudo antes de cruzar `startMovementMeters` é descartado.
        let anchor = samples.first { $0.distance >= startMovementMeters } ?? samples[0]
        let runDistance = samples.last!.distance - anchor.distance
        guard runDistance >= minTrailingMeters else { return [] }

        var splits: [KmSplit] = []
        var prev = anchor
        let fullKms = Int(runDistance / 1000.0)

        var km = 1
        while km <= fullKms {
            let boundary = anchor.distance + Double(km) * 1000.0
            let at = interpolate(samples, atDistance: boundary)
            splits.append(KmSplit(
                kilometer: km,
                startDate: prev.date,
                endDate: at.date,
                distanceMeters: 1000,
                durationSeconds: max(0, at.movingTime - prev.movingTime),
                elevationDelta: at.altitude - prev.altitude
            ))
            prev = at
            km += 1
        }

        // Km parcial final.
        let leftover = runDistance - Double(fullKms) * 1000.0
        if leftover >= minTrailingMeters {
            let last = samples.last!
            splits.append(KmSplit(
                kilometer: fullKms + 1,
                startDate: prev.date,
                endDate: last.date,
                distanceMeters: leftover,
                durationSeconds: max(0, last.movingTime - prev.movingTime),
                elevationDelta: last.altitude - prev.altitude
            ))
        }

        return splits
    }

    /// Distância real percorrida em cada intervalo de tempo, derivada da rota com o mesmo
    /// filtro de salto/pausa dos splits. Usada para dar distância (e portanto pace) real aos
    /// laps de treino — ratear a distância total pelo tempo dava a todo lap o pace médio da
    /// sessão. Retorna nil sem amostras suficientes.
    static func distances(forRanges ranges: [(start: Date, end: Date)], from locations: [CLLocation], pauses: [PauseInterval] = []) -> [Double]? {
        let samples = normalize(locations, pauses: pauses)
        guard samples.count >= 2 else { return nil }
        return ranges.map { range in
            max(0, distance(at: range.end, samples) - distance(at: range.start, samples))
        }
    }

    /// Distância acumulada (já filtrada) interpolada na data `date`.
    private static func distance(at date: Date, _ samples: [Sample]) -> Double {
        guard let first = samples.first, let last = samples.last else { return 0 }
        if date <= first.date { return first.distance }
        if date >= last.date { return last.distance }
        for i in 1..<samples.count {
            let a = samples[i - 1]
            let b = samples[i]
            if b.date >= date {
                let span = b.date.timeIntervalSince(a.date)
                guard span > 0 else { return b.distance }
                let f = date.timeIntervalSince(a.date) / span
                return a.distance + (b.distance - a.distance) * f
            }
        }
        return last.distance
    }

    // MARK: - Private

    /// Constrói a linha do tempo normalizada a partir dos fixes crus, aplicando o mesmo filtro
    /// de salto do tracker ao vivo e descontando o tempo de pausas/buracos.
    private static func normalize(_ locations: [CLLocation], pauses: [PauseInterval]) -> [Sample] {
        let sorted = locations.sorted { $0.timestamp < $1.timestamp }
        guard let first = sorted.first else { return [] }

        var samples: [Sample] = [
            Sample(date: first.timestamp, movingTime: 0, distance: 0, altitude: first.altitude)
        ]
        var lastGood = first
        var cumDistance: Double = 0
        var cumTime: Double = 0

        for i in 1..<sorted.count {
            let loc = sorted[i]
            let delta = loc.distance(from: lastGood)
            let dt = loc.timestamp.timeIntervalSince(lastGood.timestamp)
            guard dt > 0 else { continue }

            // Salto de GPS: velocidade implausível → ignora SEM avançar `lastGood`, para medir o
            // próximo ponto a partir do último bom (espelha RunTracker.processNewLocation).
            let impliedSpeed = delta / dt
            if impliedSpeed > maxPlausibleSpeed { continue }

            // O passo cruza uma pausa explícita? Então é tempo/posição "morta" entre o fim de um
            // trecho e a retomada: nem distância (a corda é deriva do GPS parado), nem tempo. Sem
            // isto o gap de uma pausa — cuja corda dá velocidade implícita ≥ 0.5 m/s — era lido
            // como blackout de GPS em movimento, creditando distância fantasma e contando o tempo
            // parado como movimento (splits não fechavam com o total). Espelha o caminho ao vivo,
            // onde os fixes durante a pausa nem são gravados.
            let crossesPause = pauses.contains { $0.start < loc.timestamp && $0.end > lastGood.timestamp }

            // Gap grande + pouco movimento → parado/pausa: não conta o tempo. Gap grande com
            // movimento real (blackout de GPS enquanto corria) → conta o tempo todo.
            let timeStep = crossesPause ? 0 : ((dt > gapCapSeconds && impliedSpeed < stationarySpeed) ? 0 : dt)

            // Gap com movimento: a corda entre o último ponto bom e este perde as curvas
            // do trajeto — contar tempo cheio com distância encurtada fazia o km do buraco
            // sair mais lento que o corrido. Credita a distância estimada pela velocidade
            // dos últimos 30s antes do gap (a corda permanece como piso; teto plausível).
            var distStep = crossesPause ? 0 : delta
            if !crossesPause,
               dt > gapCapSeconds,
               dt <= maxBridgeGapSeconds,
               impliedSpeed >= stationarySpeed,
               let speedBefore = recentSpeed(samples) {
                distStep = min(max(delta, speedBefore * dt), maxPlausibleSpeed * dt)
            }

            cumDistance += distStep
            cumTime += timeStep
            samples.append(Sample(
                date: loc.timestamp,
                movingTime: cumTime,
                distance: cumDistance,
                altitude: loc.altitude
            ))
            lastGood = loc
        }

        return samples
    }

    /// Velocidade média (m/s) nos últimos `speedLookbackSeconds` de tempo em movimento
    /// já normalizados. Nil quando não há histórico suficiente ou a velocidade não é
    /// plausível para corrida.
    private static func recentSpeed(_ samples: [Sample]) -> Double? {
        guard let last = samples.last else { return nil }
        let anchor = samples.last(where: { last.movingTime - $0.movingTime >= speedLookbackSeconds })
            ?? samples.first
        guard let anchor else { return nil }
        let timeSpan = last.movingTime - anchor.movingTime
        guard timeSpan >= 5 else { return nil }
        let speed = (last.distance - anchor.distance) / timeSpan
        guard speed >= stationarySpeed, speed <= maxPlausibleSpeed else { return nil }
        return speed
    }

    /// Interpola linearmente a amostra (tempo/data/altitude) na distância acumulada `d`.
    private static func interpolate(_ samples: [Sample], atDistance d: Double) -> Sample {
        if d <= samples.first!.distance { return samples.first! }
        if d >= samples.last!.distance { return samples.last! }

        for i in 1..<samples.count {
            let a = samples[i - 1]
            let b = samples[i]
            if b.distance >= d {
                let span = b.distance - a.distance
                guard span > 0 else { return b }
                let f = (d - a.distance) / span
                return Sample(
                    date: a.date.addingTimeInterval(b.date.timeIntervalSince(a.date) * f),
                    movingTime: a.movingTime + (b.movingTime - a.movingTime) * f,
                    distance: d,
                    altitude: a.altitude + (b.altitude - a.altitude) * f
                )
            }
        }
        return samples.last!
    }
}

File: athly-android/app/src/main/java/com/athly/runner/domain/run/SplitCalculator.kt
package com.athly.runner.domain.run

import com.athly.runner.domain.model.KmSplit
import com.athly.runner.domain.model.RoutePoint
import com.athly.runner.domain.model.distanceTo
import java.time.Instant

/**
 * Algoritmo ÚNICO de splits por km — porta exata de `SplitCalculator.swift`. Compartilhado entre o
 * tracker ao vivo ([RunTracker]) e a leitura de rota do Health Connect (prompt 12): centralizar evita
 * que os dois caminhos divirjam. Opera sobre [RoutePoint] (domínio, puro → testável) usando a MESMA
 * distância Haversine do tracker.
 *
 * Regras: salto de GPS rejeitado por plausibilidade de velocidade (≤ 7 m/s); tempo parado/buraco de GPS
 * (pausa, tela bloqueada) excluído da duração; o relógio do km 1 só começa no primeiro movimento real
 * (âncora ≥ 20 m); fronteiras de km por interpolação linear; km parcial final só com ≥ 50 m.
 */
object SplitCalculator {

    /** Velocidade máxima plausível em corrida (m/s ≈ 2:23/km). Acima disso é salto de GPS. */
    const val MAX_PLAUSIBLE_SPEED = 7.0
    /** Gap (s) entre fixes acima do qual o intervalo é candidato a pausa/blackout de GPS. */
    const val GAP_CAP_SECONDS = 6.0
    /** Velocidade (m/s) abaixo da qual um gap grande é tratado como parado (não conta tempo). */
    const val STATIONARY_SPEED = 0.5
    /** Distância (m) do "primeiro movimento": o relógio do km 1 só começa aqui. */
    const val START_MOVEMENT_METERS = 20.0
    /** Distância mínima (m) para registrar um km parcial final. */
    const val MIN_TRAILING_METERS = 50.0
    /** Gap máximo (s) que ainda recebe crédito de distância pela velocidade pré-gap. */
    const val MAX_BRIDGE_GAP_SECONDS = 90.0
    /** Janela (s de tempo em movimento) usada para estimar a velocidade pré-gap. */
    const val SPEED_LOOKBACK_SECONDS = 30.0

    private data class Sample(
        val date: Instant,
        val movingTime: Double,
        val distance: Double,
        val altitude: Double,
    )

    /** Quebra os pontos em splits de 1 km. Retorna `[]` sem dados suficientes. */
    fun kmSplits(locations: List<RoutePoint>, pauses: List<PauseInterval> = emptyList()): List<KmSplit> {
        val samples = normalize(locations, pauses)
        if (samples.size < 2) return emptyList()

        val anchor = samples.firstOrNull { it.distance >= START_MOVEMENT_METERS } ?: samples[0]
        val runDistance = samples.last().distance - anchor.distance
        if (runDistance < MIN_TRAILING_METERS) return emptyList()

        val splits = mutableListOf<KmSplit>()
        var prev = anchor
        val fullKms = (runDistance / 1000.0).toInt()

        var km = 1
        while (km <= fullKms) {
            val boundary = anchor.distance + km * 1000.0
            val at = interpolate(samples, boundary)
            splits.add(
                kmSplit(
                    kilometer = km,
                    start = prev.date,
                    end = at.date,
                    distanceMeters = 1000.0,
                    durationSeconds = maxOf(0.0, at.movingTime - prev.movingTime),
                    elevationDelta = at.altitude - prev.altitude,
                ),
            )
            prev = at
            km += 1
        }

        val leftover = runDistance - fullKms * 1000.0
        if (leftover >= MIN_TRAILING_METERS) {
            val last = samples.last()
            splits.add(
                kmSplit(
                    kilometer = fullKms + 1,
                    start = prev.date,
                    end = last.date,
                    distanceMeters = leftover,
                    durationSeconds = maxOf(0.0, last.movingTime - prev.movingTime),
                    elevationDelta = last.altitude - prev.altitude,
                ),
            )
        }

        return splits
    }

    /**
     * Distância real percorrida em cada faixa de tempo (mesmo filtro de salto/pausa) — usada para dar
     * pace real aos laps de treino (prompt 12). Retorna `null` sem amostras suficientes.
     */
    fun distances(
        ranges: List<Pair<Instant, Instant>>,
        locations: List<RoutePoint>,
        pauses: List<PauseInterval> = emptyList(),
    ): List<Double>? {
        val samples = normalize(locations, pauses)
        if (samples.size < 2) return null
        return ranges.map { (start, end) ->
            maxOf(0.0, distanceAt(end, samples) - distanceAt(start, samples))
        }
    }

    private fun kmSplit(
        kilometer: Int,
        start: Instant,
        end: Instant,
        distanceMeters: Double,
        durationSeconds: Double,
        elevationDelta: Double,
    ): KmSplit = KmSplit(
        kilometer = kilometer,
        startDate = start,
        endDate = end,
        distanceMeters = distanceMeters,
        durationSeconds = durationSeconds,
        elevationDelta = elevationDelta,
        paceSecondsPerKm = if (distanceMeters > 0) durationSeconds / (distanceMeters / 1000.0) else 0.0,
    )

    private fun normalize(locations: List<RoutePoint>, pauses: List<PauseInterval>): List<Sample> {
        val sorted = locations.sortedBy { it.timestamp }
        val first = sorted.firstOrNull() ?: return emptyList()

        val samples = mutableListOf(Sample(first.timestamp, 0.0, 0.0, first.altitude))
        var lastGood = first
        var cumDistance = 0.0
        var cumTime = 0.0

        for (i in 1 until sorted.size) {
            val loc = sorted[i]
            val delta = loc.distanceTo(lastGood)
            val dt = seconds(lastGood.timestamp, loc.timestamp)
            if (dt <= 0) continue

            // Salto de GPS: velocidade implausível → ignora SEM avançar `lastGood`.
            val impliedSpeed = delta / dt
            if (impliedSpeed > MAX_PLAUSIBLE_SPEED) continue

            val crossesPause = pauses.any { it.start < loc.timestamp && it.end > lastGood.timestamp }

            val timeStep = when {
                crossesPause -> 0.0
                dt > GAP_CAP_SECONDS && impliedSpeed < STATIONARY_SPEED -> 0.0
                else -> dt
            }

            var distStep = if (crossesPause) 0.0 else delta
            if (!crossesPause &&
                dt > GAP_CAP_SECONDS &&
                dt <= MAX_BRIDGE_GAP_SECONDS &&
                impliedSpeed >= STATIONARY_SPEED
            ) {
                recentSpeed(samples)?.let { speedBefore ->
                    distStep = minOf(maxOf(delta, speedBefore * dt), MAX_PLAUSIBLE_SPEED * dt)
                }
            }

            cumDistance += distStep
            cumTime += timeStep
            samples.add(Sample(loc.timestamp, cumTime, cumDistance, loc.altitude))
            lastGood = loc
        }

        return samples
    }

    private fun recentSpeed(samples: List<Sample>): Double? {
        val last = samples.lastOrNull() ?: return null
        val anchor = samples.lastOrNull { last.movingTime - it.movingTime >= SPEED_LOOKBACK_SECONDS }
            ?: samples.firstOrNull() ?: return null
        val timeSpan = last.movingTime - anchor.movingTime
        if (timeSpan < 5) return null
        val speed = (last.distance - anchor.distance) / timeSpan
        if (speed < STATIONARY_SPEED || speed > MAX_PLAUSIBLE_SPEED) return null
        return speed
    }

    private fun interpolate(samples: List<Sample>, distance: Double): Sample {
        if (distance <= samples.first().distance) return samples.first()
        if (distance >= samples.last().distance) return samples.last()

        for (i in 1 until samples.size) {
            val a = samples[i - 1]
            val b = samples[i]
            if (b.distance >= distance) {
                val span = b.distance - a.distance
                if (span <= 0) return b
                val f = (distance - a.distance) / span
                return Sample(
                    date = lerpInstant(a.date, b.date, f),
                    movingTime = a.movingTime + (b.movingTime - a.movingTime) * f,
                    distance = distance,
                    altitude = a.altitude + (b.altitude - a.altitude) * f,
                )
            }
        }
        return samples.last()
    }

    private fun distanceAt(date: Instant, samples: List<Sample>): Double {
        val first = samples.first()
        val last = samples.last()
        if (date <= first.date) return first.distance
        if (date >= last.date) return last.distance
        for (i in 1 until samples.size) {
            val a = samples[i - 1]
            val b = samples[i]
            if (b.date >= date) {
                val span = seconds(a.date, b.date)
                if (span <= 0) return b.distance
                val f = seconds(a.date, date) / span
                return a.distance + (b.distance - a.distance) * f
            }
        }
        return last.distance
    }

    private fun seconds(from: Instant, to: Instant): Double =
        (to.toEpochMilli() - from.toEpochMilli()) / 1000.0

    private fun lerpInstant(a: Instant, b: Instant, f: Double): Instant =
        Instant.ofEpochMilli((a.toEpochMilli() + (b.toEpochMilli() - a.toEpochMilli()) * f).toLong())
}
